pkg/streamingpromql/operators/aggregations/limitklimitratio/operator.go

```diff
@@ -0,0 +1,409 @@
+	annotations        *annotations.Annotations
+
+	// index by output series id - populated via the SeriesMetadata() and then returned in the NextSeries()
+	values []*types.InstantVectorSeriesData
+
+	// the next output series to be returned in NextSeries()
+	nextSeriesIndex int
+
+	// true if this is a limit_ratio operation
+	isRatio bool
+}
+
+// isLimitK returns true if this is a limitk operation.
+// This is helper function to avoid negatives in limitk related conditionals.
+func (t *Operator) isLimitK() bool {
+	return !t.isRatio
+}
+
+// initParam validates the given parameter and determines the k or ratio for each step.
+func (t *Operator) initParam(ctx context.Context, annotations *annotations.Annotations, memoryConsumptionTracker *limiter.MemoryConsumptionTracker, stepCount int, param types.ScalarOperator) error {
+	var err error
+
+	if t.isRatio {
+		t.stepArg, err = newRatioParam(ctx, annotations, memoryConsumptionTracker, stepCount, param)
+	} else {
+		t.stepArg, err = newKParam(ctx, memoryConsumptionTracker, stepCount, param)
+	}
+	return err
+}
+
+func (t *Operator) SeriesMetadata(ctx context.Context, matchers types.Matchers) ([]types.SeriesMetadata, error) {
+	var err error
+
+	// parse and validate the given input argument for this aggregate (ie k or ratio)
+	if err = t.initParam(ctx, t.annotations, t.MemoryConsumptionTracker, t.TimeRange.StepCount, t.Param); err != nil {
+		return nil, err
+	}
+	defer t.stepArg.close()
+
+	// all the k/ratio values are 0 so we will have nothing to return
+	if t.stepArg.allZero() {
+		return nil, nil
+	}
+
+	// all the series metadata we should consider
+	innerSeries, err := t.Inner.SeriesMetadata(ctx, matchers)
+	if err != nil {
+		return nil, err
+	}
+	defer types.SeriesMetadataSlicePool.Put(&innerSeries, t.MemoryConsumptionTracker)
+
+	groupLabelsBytesFunc := aggregations.GroupLabelsBytesFunc(t.Grouping, t.Without)
+	groups := map[string]*limitGroup{}
+	seriesToGroups := make([]*limitGroup, len(innerSeries))
+
+	defer func() {
+		for _, group := range seriesToGroups {
+			group.close()
+		}
+	}()
+
+	// Go through each series and find / create its group, and keep track of how many series contribute to each group.
+	// We do this separately to the loop below so that we know the number of series in each group when we allocate
+	// each group's `series` slice inside accumulateValue - this allows us to avoid allocating a huge slice if the
+	// group only has a few series and `k` is large.
+	for idx, series := range innerSeries {
+		groupLabelsString := groupLabelsBytesFunc(series.Labels)
+		g, groupExists := groups[string(groupLabelsString)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !groupExists {
+			// note - only limitk uses the step counter
+			g, err = newLimitGroup(t.TimeRange.StepCount, t.MemoryConsumptionTracker, t.isLimitK())
+			if err != nil {
+				// in the case of an early exit all the groups can be closed
```

Comment:
[nit] Not sure this comment is adding much - we defer closing all the groups above, so won't all groups be closed regardless of the outcome of this method?